lib/queue.js

```diff
@@ -1,256 +1,497 @@
+    client.multi()
+      .llen(this.toKey('waiting'))
+      .llen(this.toKey('active'))
+      .scard(this.toKey('succeeded'))
+      .scard(this.toKey('failed'))
+      .zcard(this.toKey('delayed'))
+      .get(this.toKey('id'))
+      .exec(multi.defer());
+    return multi;
+  }).then((results) => ({
+    waiting: results[0],
+    active: results[1],
+    succeeded: results[2],
+    failed: results[3],
+    delayed: results[4],
+    newestJob: results[5] ? parseInt(results[5], 10) : 0
+  }));
+
+  if (cb) helpers.asCallback(promise, cb);
+  return promise;
+};
+
+Queue.prototype._scanForJobs = function (key, cursor, size, set, cb) {
+  const batchCount = Math.min(size, this.settings.redisScanCount);
+  this.client.sscan(key, cursor, 'COUNT', batchCount, (err, results) => {
+    /* istanbul ignore if */
+    if (err) {
+      return cb(err);
+    }
+
+    const nextCursor = results[0];
+    const ids = results[1];
+
+    // A given element may be returned multiple times in SSCAN.
+    // So, we use a set to remove duplicates.
+    // https://redis.io/commands/scan#scan-guarantees
+    for (let id of ids) {
+      // For small sets, encoded as intsets, SSCAN will ignore COUNT.
+      // https://redis.io/commands/scan#the-count-option
+      if (set.size === size) break;
+
+      set.add(id);
+    }
+
+    if (nextCursor === '0' || set.size >= size) {
+      return cb(null, set);
+    }
+
+    this._scanForJobs(key, nextCursor, size, set, cb);
+  });
+};
+
+/**
+ * Get jobs from queue type.
+ *
+ * @param {String} type The queue type ("failed", "succeeded", "waiting", etc.)
+ * @param {Object} page
+ * @param {Number=} page.start Start of query range for waiting/active/delayed
+ *   queue types
+ * @param {Number=} page.end End of query range for waiting/active/delayed queue
+ *   types
+ * @param {Number=} page.size Number jobs to return for failed/succeeded (SET)
+ *   types
+ * @param {Function=} callback Called with the equivalent of the returned
+ *   promise.
+ * @return {Promise<Job[]>} Resolves to the jobs the function found.
+ */
+Queue.prototype.getJobs = function (type, page, cb) {
+  const promise = this._commandable().then((client) => {
+    const idsPromise = helpers.deferred(), next = idsPromise.defer();
+    switch (type) {
+      case 'failed':
+      case 'succeeded':
+        this._scanForJobs(this.toKey(type), '0', page.size, new Set(), next);
+        break;
+      case 'waiting':
+      case 'active':
+        client.lrange(this.toKey(type), page.start, page.end, next);
+        break;
+      case 'delayed':
+        client.zrange(this.toKey(type), page.start, page.end, next);
+        break;
+      default:
+        throw new Error('Improper queue type');
+    }
+
+    return idsPromise;
+  }).then((ids) => {
+    // If ids is a Set (i.e. returned from Queue._scanForJobs), convert it to an
+    // Array.
+    const idArray = Array.from(ids);
+    return Promise.all(idArray.map((id) => this.getJob(id)));
```

Comment:
This will end up doing a bunch of `hget`s in `fromId` for jobs that aren't stored in the local jobs map; we *could* optimize by doing an `hmget` if we were concerned about getJobs performance. Since we don't use it anywhere else inside bee-queue and it's mainly provided for inspection by e.g. arena, I'm not too worried about this, but with large pages of delayed jobs, it might be an appreciable performance difference.